docs/devel/api_changes.md

```diff
@@ -0,0 +1,237 @@
+all API changes will need all of these steps.
+
+## On compatibility
+
+Before talking about API changes, it is worthwhile to clarify what we mean by
+API compatibility.  An API change is considered compatible if it:
+   * adds new functionality that is not required for correct behavior
+   * does not change existing semantics
+   * does not change existing defaults
+
+Put another way:
+
+1. An API call (e.g. a structure POSTed to a REST endpoint) that worked before
+   your change must work the same after your change.
+2. An API call that uses your change must work correctly against servers that
+   do not include your change.
+3. It must be possible to round-trip your change (convert to different API
+   versions and back) with no loss of information.
+
+If your change does not meet these criteria, it is not considered strictly
+compatible.  There are times when this might be OK, but mostly we want changes
+that meet this definition.
+
+Let's consider some examples.  In a hypothetical API (assume we're at version
+v6), the `Frobber` struct looks something like this:
+
+```go
+// API v6.
+type Frobber struct {
+	Height int    `json:"height"`
+	Param  string `json:"param"`
+}
+```
+
+You want to add a new `Width` field.  It is generally safe to add new fields
+without changing the API version, so you can simply change it to:
+
+```go
+// Still API v6.
+type Frobber struct {
+	Height int    `json:"height"`
+	Width  int    `json:"width"`
+	Param  string `json:"param"`
+}
+```
+
+The onus is on you to define a sane default value for `Width` such that rule #1
+above is true - API calls that used to work must continue to work.
+
+For your next change you want to allow multiple `Param` values.  You can not
+simply change `Param string` to `Params []string` without reviing the whole API
+- that fails rules #1 and #2.  You can instead do something like:
+
+```go
+// Still API v6, but kind of clumsy.
+type Frobber struct {
+	Height int           `json:"height"`
+	Width  int           `json:"width"`
+	Param  string        `json:"param"`  // the first param
+	ExtraParams []string `json:"params"` // additional params
+}
+```
+
+Now you can satisfy the rules: API calls that provide the old style `Param`
+will still work, while servers that don't understand `ExtraParams` can ignore
+it.  This is somewhat unsatisfying as an API, but it is strictly compatible.
+
+Part of the reason for versioning APIs and for using an explicit internal API
+is to handle growth like this.  The internal API can be implemented as:
+
+```go
+// Internal API, soon to be v7beta1.
+type Frobber struct {
+	Height int           `json:"height"`
+	Width  int           `json:"width"`
+	Params []string      `json:"params"`
+}
+```
+
+The code that converts to/from versioned APIs can decode this into the somewhat
+uglier (but compatible!) structures.  Eventually, a new API version, let's call
+it v7beta1, will be forked and it can use the internal structure.
```

Comment:
Conversely, if we create a new, clean versioned API representation, we generally change the internal representation to match, and write conversions for the older API versions.